### ISSUE TITLE
Be able to listen to hooks outside of main Module file

### ISSUE THREAD

**Issue by matks** [PrestaShop/PrestaShop#14854]
Currently, modules which want to plug themselves on a PrestaShop hook must do so using `registerHook(...)` function and implement the right function in the main Module class file (= the one that extends `Module`) 

This obliges developer to put all the hook behaviors in a single class, preventing them to split the code into multiple files, test them independently or use PSR4 autoloading with namespace.

How could we change that ?

Idea 1: creates an a php trait `Hookable` which allows these behaviors
Idea 2: creates an interface `Hookable` and require the module developer to list the hookable classes into a YAML / XML configuration file

**Comment by khouloudbelguith** [comment]
Ping @PrestaShop/prestashop-core-developers

**Comment by PierreRambaud** [comment]
The code can currently be splitted as we do for facetedsearch (even if it's not the better way to do it).
As Doctrine do for Entity. Can we create `modules/my-module/src/Hooks/HookName.php` in addition?

**Comment by jolelievre** [comment]
I started a quick proposal of code, I prefer to copy it here before I lose from my VSCode cache ^^
This was mainly designed with DataDomain hooks but I think the idea of a Hookable trait is more generic and very usefull as well. We could propose a specific domain entity trait based on the Hookable one.

```php
class DomainDataHookHandler {
    private $domainEntity;

    public function __construct($domainEntity) {
        $this->domainEntity = $domainEntity;
    }

    public function getDomainEntity() {
        return $this->domainEntity;
    }

    public function getHooks() {
        return [
            'hookAction' . $this->domainEntity .'GridDefinitionModifier',
            'hookAction' . $this->domainEntity .'GridQueryBuilderModifier',
            'hookAction' . $this->domainEntity .'FormBuilderModifier',
        ];
    }

    public function hookActionGridDefinitionModifier(array $params) {
        //Override this method to manage grid definition
    }

    public function hookActionGridQueryBuilderModifier(array $params) {
        //Override this method to manage grid query builder
    }

    public function hookActionFormBuilderModifier(array $params) {
        //Override this method to manage form builder
    }

    ....
}

class CustomerDomainDataHookHandler()
{
    public function __construct() {
        parent::__construct('Customer');
    }
}

class ModuleCore {
    private $domainHandlers = [
        'Customer' => DomainDataHookHandler::class,
    ];

    public function install() {
        foreach ($this->domainHandlers as $domainHandlerClass) {
            $domainHandler = new $domainHandlerClass();
            foreach ($domainHandler->getHooks() as $hooName) {
                $this->registerHook($hookName);
            }
        }

        return parent::install();
    }

    public function __call($method, $arguments) {
        $domainEntity = $this->getDomainEntity($method);
        if (isset($this->domainHandlers[$domainEntity])) {
            $domainHandler = new $this->domainHandlers[$domainEntity];
            $hookMethod = $this->getHookMethod($method);
            $domainHandler->$hookMethod($arguments);
        }
    }
}```

**Comment by jolelievre** [comment]
As @PierreRambaud suggested we could also use a folder as convention to store these Hookables. But @Quetzacoalt91 also suggested to use services which would be very interesting, this would allow Hookable classes to have dependencies as well (Translator, Doctrine, ...)
And maybe introduce a new tag to identify and register these services automatically

**Comment by eternoendless** [comment]
Why not use events or pseudo events instead of creating a new thing?

1. Module declares event subscribers
2. Hook dispatcher dispatches the event (and either still tries to call other modules not using events OR a there's a DefaultSubscriber that takes care of that)
3. Profit

Yes, the implementation won't be as easy as 1-2-3 but from a developer point of view it sounds enticing.

**Comment by matks** [comment]
@Quetzacoalt91 I noticed this `classes/Hook/HookInterface.php` in ps_googleanalytics v4.0.0

https://github.com/PrestaShop/ps_googleanalytics/

Is this the beginning of an answer 😄  ?

**Comment by Quetzacoalt91** [comment]
Exactly, and this would be even better if the relation between a triggered hook and the class to call could be set by the services definition file (like we do with SF events).

**Comment by LouiseBonnard** [comment]
Hi @PrestaShop/prestashop-core-developers, can this issue be moved to the backlog?

**Comment by PierreRambaud** [comment]
@LouiseBonnard Not yet. We must decide what we will do about this issue :)
